electricitymap/contrib/libs/models/datapoints.py

```diff
@@ -0,0 +1,427 @@
+        datetime: datetime,
+        source: str,
+        consumption: float,
+        forecasted: bool = False,
+    ) -> Optional["Consumption"]:
+        try:
+            return Consumption(
+                zoneKey=zoneKey,
+                datetime=datetime,
+                source=source,
+                consumption=consumption,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(
+                f"Error creating consumption Event {datetime}: {e}",
+                extra={
+                    "zoneKey": zoneKey,
+                    "datetime": datetime,
+                    "kind": "consumption",
+                },
+            )
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "consumption": self.consumption,
+            "source": self.source,
+        }
+
+
+class Price(Event):
+    price: float
+    currency: str
+
+    @validator("currency")
+    def _validate_currency(cls, v: str):
+        if v not in VALID_CURRENCIES:
+            raise ValueError(f"Unknown currency: {v}")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zoneKey: ZoneKey,
+        datetime: datetime,
+        source: str,
+        price: float,
+        currency: str,
+        forecasted: bool = False,
+    ) -> Optional["Price"]:
+        try:
+            return Price(
+                zoneKey=zoneKey,
+                datetime=datetime,
+                source=source,
+                price=price,
+                currency=currency,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating price Event {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "currency": self.currency,
+            "price": self.price,
+            "source": self.source,
+        }
+
+
+class EventList(ABC):
+    """A wrapper around Events lists."""
+
+    logger: Logger
+    events: List[Event]
+
+    def __init__(self, logger: Logger):
+        self.events = list()
+        self.logger = logger
+
+    @abstractmethod
+    def append(self, **kwargs):
+        """Handles creation of events and adding it to the batch."""
+        # TODO Handle one day the creation of mixed batches.
+        pass
+
+    def to_list(self) -> List[Dict[str, Any]]:
+        return [event.to_dict() for event in self.events]
+
+
+class ExchangeList(EventList):
+    def append(
+        self,
+        zoneKey: ZoneKey,
+        datetime: datetime,
+        source: str,
+        value: float,
+        forecasted: bool = False,
+    ):
+        event = Exchange.create(
+            self.logger, zoneKey, datetime, source, value, forecasted
+        )
+        if event:
+            self.events.append(event)
```

Comment:
We have a part in the backend that checks for duplicates before inserting in the database. In a V2 we could move this check here and use it to remove duplicates. Because at the moment if a batch contains duplicates we raise an error and discard it.